cmd/akamai-purger/main.go

```diff
@@ -193,36 +194,53 @@ func (ap *akamaiPurger) len() int {
 
 func (ap *akamaiPurger) purge() error {
 	ap.Lock()
-	queueEntries := ap.toPurge[:]
-	ap.toPurge = [][]string{}
-	ap.Unlock()
-	if len(queueEntries) == 0 {
+	queueSize := len(ap.toPurge)
+	batchSize := ap.queueEntriesPerBatch
+
+	// If the queue is empty, return immediately.
+	if queueSize <= 0 {
+		ap.Unlock()
 		return nil
 	}
 
-	stoppedAt, err := ap.client.Purge(queueEntries)
+	// If the queue contains less than full batch, set the batch size to the
+	// queue size.
+	if queueSize < batchSize {
+		batchSize = queueSize
+	}
+	batchBegin := queueSize - batchSize
+	purgeBatch := ap.toPurge[batchBegin:]
+	ap.toPurge = ap.toPurge[:batchBegin]
+	ap.Unlock()
+
+	err := ap.client.Purge(purgeBatch)
 	if err != nil {
 		ap.Lock()
-
-		// Add the remaining queue entries back, but at the end of the queue. If somehow
-		// there's a URL which repeatedly results in error, it won't block the
-		// entire queue, only a single batch.
-		ap.toPurge = append(ap.toPurge, queueEntries[stoppedAt:]...)
+		if len(ap.toPurge) >= ap.maxQueueSize {
+			// Drop the oldest entry from the queue to make room.
+			ap.toPurge = ap.toPurge[batchSize:]
+		}
+		// Add the entries from the failed batch to the back of the queue to
+		// avoid blocking the entire queue.
+		ap.toPurge = append(purgeBatch, ap.toPurge...)
+		ap.log.Errf("Failed to purge OCSP responses for %d certificates: %s", queueSize, err)
 		ap.Unlock()
-		ap.log.Errf("Failed to purge OCSP responses for %d certificates: %s", len(queueEntries), err)
 		return err
 	}
 	return nil
 }
 
-// Purge is an exported gRPC method which receives purge requests and appends
-// them to the queue.
+// Purge is an exported gRPC method which receives purge requests containing
+// URLs and prepends them to the purger queue.
 func (ap *akamaiPurger) Purge(ctx context.Context, req *akamaipb.PurgeRequest) (*emptypb.Empty, error) {
 	ap.Lock()
 	defer ap.Unlock()
-	if len(ap.toPurge) >= ap.maxQueueSize {
-		return nil, errors.New("akamai-purger queue too large")
+	queueSize := len(ap.toPurge)
+	if queueSize >= ap.maxQueueSize {
+		// Drop the oldest entry from the queue to make room.
+		ap.toPurge = ap.toPurge[1:]
 	}
+	// Add the entry from the new request to the front of the queue.
```

Comment:
Here's another interesting/surprising allocation issue: `ap.toPurge = ap.toPurge[1:]` is cheap, because we're just subslicing a slice. No copying or allocation can possibly happen here.

However, immediately below, we _append_ to `toPurge`. That can cause a reallocation (which includes a copy of all elements).

How often will it cause a reallocation? It depends on how much extra space is in `toPurge`'s backing array. If `len(toPurge) == 1000000` and `cap(toPurge) == 2000000`, we can "drop the oldest entry and add a new entry" 1M times before we have to reallocate.

If `len(toPurge) == 1000000` and `cap(toPurge) == 1000000`, we will immediately allocate when we add a new entry. That's because dropping an entry didn't actually make any room in the backing array, it just moved the beginning of the slice forward by one element, and decreased the slice's length.

So, when we're maxed out, are we going to be reallocating every time we push to the stack? I think no. When append reallocates a slice's backing array, it generally allocates extra wiggle room proportional to the slice's size. I think in this case it winds up being about 25% extra room. So if the stack has size 1M, we can "drop oldest; push new" 250k times before we have to reallocate. But I'd have to double check the `append` implementation to be sure.

No change needed here.

Reply:
So, modeled this a little:

```
      0:        0x0     0
      2: 0x1400012c020 1        ->      2       200%
      3: 0x1400012e040 2        ->      4       200%
      5: 0x14000124080 4        ->      8       200%
      9: 0x14000132000 8        ->      16      200%
     17: 0x14000134000 16       ->      32      200%
     33: 0x14000136000 32       ->      64      200%
     65: 0x14000138000 64       ->      128     200%
    129: 0x1400013a000 128      ->      256     200%
    257: 0x1400013c000 256      ->      512     200%
    513: 0x1400013e000 512      ->      848     165%
    849: 0x14000148000 848      ->      1280    150%
   1281: 0x14000152000 1280     ->      1792    140%
   1793: 0x14000160000 1792     ->      2560    142%
   2561: 0x1400016a000 2560     ->      3408    133%
   3409: 0x14000180000 3408     ->      5120    150%
   5121: 0x1400018a000 5120     ->      7168    140%
   7169: 0x14000198000 7168     ->      9216    128%
   9217: 0x140001aa000 9216     ->      12288   133%
  12289: 0x140001c2000 12288    ->      16384   133%
  16385: 0x140001e2000 16384    ->      21504   131%
  21505: 0x1400020c000 21504    ->      27648   128%
  27649: 0x14000242000 27648    ->      34816   125%
  34817: 0x14000286000 34816    ->      44032   126%
  44033: 0x140002dc000 44032    ->      55296   125%
  55297: 0x14000348000 55296    ->      69632   125%
  69633: 0x140003d0000 69632    ->      88064   126%
  88065: 0x1400047c000 88064    ->      110592  125%
 110593: 0x14000180000 110592   ->      139264  125%
 139265: 0x14000600000 139264   ->      175104  125%
 175105: 0x14000290000 175104   ->      219136  125%
 219137: 0x14000756000 219136   ->      274432  125%
 274433: 0x1400096e000 274432   ->      344064  125%
 344065: 0x14000c0e000 344064   ->      431104  125%
 431105: 0x14000f58000 431104   ->      539648  125%
 539649: 0x14001376000 539648   ->      674816  125%
 674817: 0x14000600000 674816   ->      843776  125%
 843777: 0x1400189c000 843776   ->      1055744 125%
1055745: 0x14000c70000 1055744  ->      1319936 125%
```

Reply:
So it seems that we're always starting with 100% - 200% overhead each time the slice is grown.